src/util.h

```diff
@@ -112,6 +108,64 @@ static inline bool error(const char* format)
     return false;
 }
 
+/* One of these structures is statically allocated at the site of every call to
+ * LogPrint(category, format, ...), and also at the site of every call to
+ * LogAcceptCategory(category). The first time a site is reached, its structure
+ * is initialized and linked into the global linked list logFastReached (this
+ * "discovery" event is the first time it's possible to do this).
+ */
+struct LogFastEntry {
+    const char* category;
+    const char* line;   // pathname:lineNumber
+    bool enabled;
+    LogFastEntry *next;
+};
+
+/* This function is called the first time (and only the first time) a call
+ * site to LogPrint() or LogAcceptCategory() is visited.
+ */
+void LogFastEntryInit(LogFastEntry *lfe, const char *category, const char* line);
+
+/**
+ * Print the given string (with arguments) if the category is enabled, or if
+ * the line (where the LogPrint occurs) is enabled. This is very efficient
+ * except possibly for the first call in the life of the process.
+ */
+#define LogPrint(category, ...) do {                            \
+    static_assert((category) != nullptr);                       \
+    static LogFastEntry lfe;                                    \
+    static boost::once_flag once = BOOST_ONCE_INIT;             \
```

Comment:
I didn't think it's worthy of a comment, but the reason we're using `call_once()` here instead of just using a simple static `bool` is for multithread protection. I hope, and expect, that `call_once()` is efficient; it seems to me what it should do is test an `init` boolean, and if false, acquire a mutex and test `init` again (in case another thread beat us to it). (If `init` is not set, then call `LogFastEntryInit()` and *then* set `init` to true and release the mutex.) After initialization occurs, the test will be very fast (test `init` without a mutex, and since it's set, skip the initialization stuff).